File: brainiac.go
// brainiac.go                                                 
package main                                             
import "fmt"                                                         
import "os"                                             
import "bufio"                                                         
import "time"                                              
                                                         
const (                                                  
                                                         
  PYRAMIDAL = 0                                            
  INHIBITOR = 1                                            
  SPD = 40  // Synapses per Dendrite                       
  DSZ = 40 * 4 // 32 bits per Synapse                      
  PPP = 32  // Proximal per Pyramidal                      
  BPP = 32  // Basal per Pyramidal                         
  APP = 32  // Apical per Pyramidal                        
  DPN = 32 + 32 + 32                                       
  PinL2 = 28  // Pyramidal in L2                           
  Acc2  = 28                                               
  PinL4 = 28  // Pyramidal in L4                           
  Acc4  = 28 + PinL2                                       
  PinL5 = 28  // Pyramidal in L5                           
  Acc5  = 28 + Acc4                                        
  PinL6 = 28  // Pyramidal in L6                           
  Acc6  = 28 + Acc5                                        
  PinTh = 4   // Pyramidal in Thalmus                      
  AccTh = 4  + Acc6                                        
  MCinC = 96  // Minicolumns in a Column                   
  CinP = 9    // Columns in a Patch                    
  NP = 100224                                              
)                                                        
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
type Potential struct { excited bool }                                                         
                                                         
                                                         
                                                         
type Synapse int32                                       
                                                         
                                                         
                                                         
type Dendrite   struct {                                                                                    
        S [SPD]Synapse                                   
     }                                                   
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
type Neuron    struct {                                                                                     
       State int16                                                                                          
       Kind  int16                                                                                          
       De    [DPN]Dendrite                                                                                  
                                                                                                            
     }                                                                                                      
                                                         
                                                         
                                                         
                                                         
                                                        
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
type Minicolumn struct {                                                                                    
        N [AccTh]Neuron                                                          
                                                                                 
                                                                                  
                                                                                   
                                                     
     }                                                   
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
type Column struct {                                                                                    
       MC [MCinC]Minicolumn                                                       
     }                                                      
                                                         
                                                        
                                                        
                                                        
                                                        
                                                        
                                                        
                                                        
type Patch  struct {                                                                                    
       C [CinP]Column                                                           
     }                                                      
                                                         
                                                              
                              
                                                        
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
func la2sa( a int )(c,mc,ni,lv,lvo int){                                                                                 
    ni =a % AccTh                                                                                                  
   mco:= a / AccTh;       mc = mco % MCinC                                                                       
   co:= a / (AccTh*MCinC); c= co % CinP                                                                          
   if        ni<PinL2{   lv=0;   lvo= ni                                                                          
   }else if  ni<Acc4{    lv=1;   lvo= ni-PinL2                                                                    
   }else if  ni<Acc5{    lv=2;   lvo= ni-Acc4                                                                     
   }else if  ni<Acc6{    lv=3;   lvo= ni-Acc5                                                                     
   }else if  ni<AccTh{   lv=4;   lvo= ni-Acc6                                                                     
   }else{                lv=-1;  lvo=-1  }                                                                        
   return c, mc, ni, lv, lvo                                                                                          
}                                                        
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
func initialize() *Patch {                                    
                                                      
  P:=new(Patch)                                                        
  tvs:=time.Now().UnixMilli()                           
  if len(os.Args) > 1 {
    fmt.Println("  initializing", os.Args[1])            
    pexfn:="b-"+os.Args[1]+
         "-9-96-4-28-28-28-28-0-0-0.pex"; 
    dndfn:="b-"+os.Args[1]+
         "-9-96-4-28-28-28-28-32-32-32-40-0-0-0.dnd";
    fpex, err1 := os.Open(pexfn)
    if err1 == nil {
      fdnd, err2 := os.Open(dndfn)
      if err2 == nil {
        dp:= SPD+PPP+BPP+APP                           
        fmt.Println("   dendrites:   ", dp*NP)         
        fmt.Println("   p cells  :     ", NP)                
        fmt.Println("   miniclmns:        ", CinP*MCinC)      
        fmt.Println("   loading...")                         
        d    := make([]byte, DPN*SPD*4)                                  
        drdr := bufio.NewReader(fdnd)               
        fpex.Close()                         
        b,_ :=os.ReadFile(pexfn)               
        for i:=0; i<NP;i++{                            
            c,mc,ni,_,_:=la2sa(i)                      
      //  fmt.Println("--",i,c,mc,lv,lvo)    
          P.C[c].MC[mc].N[ni].Kind=                                                       
              (int16)(b[i*4+0]+(b[i*4+1]<<8))
          P.C[c].MC[mc].N[ni].State=                                       
              (int16)(b[i*4+2]+(b[i*4+3]<<8))                                          
          _,_   = drdr.Read(d    )                                                     
          for j:=0; j<DPN; j++{                                                                                  
            for k:=0; k<SPD; k++{                                                     
              P.C[c].MC[mc].N[ni].De[j].S[k]=              
              (Synapse)(d[((j*SPD)+k)*4+0]+(d[((j*SPD)+k)*4+1]<<8)+              
              (d[((j*SPD)+k)*4+2]<<16)+(d[((j*SPD)+k)*4+3]<<24))             
            }                                            
          }                                              
        }                                                
        fdnd.Close()                                     
      }else{                                             
        fmt.Println("  file",dndfn,"?",err2)             
      }                                                  
    }else{                                               
      fmt.Println("  file",pexfn,"?",err1)               
    }                                                    
    fmt.Println("  MSEC INIT:",time.Now().UnixMilli()-tvs)                       
  }else{                                                                                             
    fmt.Println("  parameter?")                          
  }                                                      
  return P                                               
}                                                        
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
func iterate(P *Patch){                                       
                                                    
  fmt.Println("  iterating")                     
                                                         
}                                             
                                                         
func summarize(P *Patch){                                     
                                                    
  fmt.Println("  summarizing")                   
                                                         
}                                         
                                                         
func main(){                                                    
  fmt.Println(" Brainiac")                       
                                                         
  P:=initialize()                                            
  iterate(P)                                           
  summarize(P)                                             
                                                         
                                                         
                                                         
                                                         
                                                         
}                                            
                                                         
                                                         
                                                         
                                                         

File: Brainiac.java
// Brainiac.java                                         
import java.io.*;                                        
import java.time.*;                                      
public class Brainiac {                                  
                                                         
                                                         
                                                         
                                                         
                                                         
  final static int PYRAMIDAL = 0;                        
  final static int INHIBITOR = 1;                        
  final static int SPD = 40; // Synapses per Dendrite    
  final static int DSZ = 40 * 4; // 32 bits per Synapse  
  final static int PPP = 32; // Proximal per Pyramidal   
  final static int BPP = 32; // Basal per Pyramidal      
  final static int APP = 32; // Apical per Pyramidal     
  final static int DPN = 32 + 32 + 32;                   
  final static int PinL2 = 28; // Pyramidal in L2        
  final static int Acc2  = 28;                           
  final static int PinL4 = 28; // Pyramidal in L4        
  final static int Acc4  = 28 + PinL2;                   
  final static int PinL5 = 28; // Pyramidal in L5        
  final static int Acc5  = 28 + Acc4;                    
  final static int PinL6 = 28; // Pyramidal in L6        
  final static int Acc6  = 28 + Acc5;                    
  final static int PinTh = 4;  // Pyramidal in Thalmus   
  final static int AccTh = 4  + Acc6;                    
  final static int MCinC = 96 ; // Minicolumns per Column
  final static int CinP = 9  ; // Columns in a Patch     
  final static int NP = 100224;                          
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
  public static class Dendrite{                          
    int[]   S;                                           
                                                         
    public Dendrite(InputStream fdnd){                   
           S = new int[SPD];                             
      byte[] b = new byte[SPD*4];                        
      try {                                              
        fdnd.read(b);                                    
        for(int i=0;i<SPD;i++){                          
          S[i]=b[i*4]+(b[i*4+1]<<8)+                     
              (b[i*4+2]<<16)+(b[i*4+3]<<24);             
        }                                                
      } catch (IOException ex){                          
        ex.printStackTrace();                            
      }                                                  
    }                                                    
  }                                                      
                                                         
  public static class   Neuron{                          
    int State;                                           
    int Kind;                                            
    Dendrite[] De;                                       
                                                         
                                                         
    public Neuron(InputStream fpex, InputStream fdnd){   
      byte[] b      = new byte[4];                       
      try                  {                             
        fpex.read(b);                                    
        De    = new Dendrite[DPN];                       
        State = b[0]+(b[1]<<8);                          
        Kind = b[2]+(b[3]<<16);                          
        for(int i=0;i<DPN;i++){                          
          De[i] = new Dendrite(fdnd);                    
        }                                                
      } catch (IOException ex) {                         
        ex.printStackTrace();                            
      }                                                  
    }                                                    
  }                                                      
                                                         
  public static class Minicolumn{                        
    Neuron[] N;                                          
                                                         
   public Minicolumn(InputStream fpex, InputStream fdnd){
            N = new    Neuron[AccTh];                    
      for(int i=0;i<AccTh;i++){                          
        N[i] = new Neuron(fpex,fdnd);                    
      }                                                  
    }                                                    
  }                                                      
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
  public static class Column{                            
    Minicolumn[] MC;                                     
                                                         
    public Column(InputStream fpex, InputStream fdnd){   
       MC = new Minicolumn[MCinC];                       
       for(int i=0;i<MCinC;i++){                         
         MC[i] = new Minicolumn(fpex,fdnd);              
       }                                                 
    }                                                    
  }                                                      
                                                         
  public static class Patch{                             
    Column[] C;                                          
                                                         
    public Patch(InputStream fpex, InputStream fdnd){    
      C = new Column[CinP];                              
      for(int i=0;i<CinP;i++){                           
        C[i] = new Column(fpex,fdnd);                    
      }                                                  
    }                                                    
  }                                                      
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
  public static int[] la2sa( int a){                     
                                                         
    int r[] = new int[5]; //c,mc,ni,lv,lvo               
                             int mci = a  % AccTh;       
    int mco = a  / AccTh;r[1] = mco % MCinC;             
    int co = a / (AccTh*MCinC);r[0]=co % CinP;           
    if      (mci<PinL2) {r[3]=0;  r[4]=mci;              
    }else if(mci<Acc4)  {r[3]=1;  r[4]=mci-PinL2;        
    }else if(mci<Acc5)  {r[3]=2;  r[4]=mci-Acc4;         
    }else if(mci<Acc6)  {r[3]=3;  r[4]=mci-Acc5;         
    }else if(mci<AccTh) {r[3]=4;  r[4]=mci-Acc6;         
    }else               {r[3]=-1; r[4]=-1;               
    }                                                    
    r[2]=mci;                                            
                                                         
                                                         
                                                         
                                                         
    return r;                                            
  }                                                      
                                                         
  public static Patch initialize(String[] args){         
    int dp,c,mc,ni,lv,lvo; int[] a;                      
                                                         
    Patch P;                                             
    Instant tvs = Instant.now();                         
    P = null;                                            
    if (args.length > 0) {                               
      System.out.println("  initializing "+args[0]);     
      String pexfn="b-"+args[0]+                         
         "-9-96-4-28-28-28-28-0-0-0.pex";                
      String dndfn="b-"+args[0]+                         
         "-9-96-4-28-28-28-28-32-32-32-40-0-0-0.dnd";    
      try (                                              
        InputStream fpex = new FileInputStream(pexfn);   
        InputStream fdnd = new FileInputStream(dndfn);   
      ) {                                                
        dp= SPD+PPP+BPP+APP;                             
        System.out.println("   dendrites:    "+dp*NP);   
        System.out.println("   p cells  :      "+NP);    
        System.out.println("   miniclmns:         "+     
                           CinP*MCinC);                  
        System.out.println("   loading...");             
        P = new Patch(fpex,fdnd);                        
        byte[] b      = new byte[NP*4];                  
        fpex.read(b);                                    
        for(int i=0;i<NP;i++){                           
    a = la2sa(i);c=a[0];mc=a[1];ni=a[2];lv=a[3];lvo=a[4];
//System.out.println("--"+i+" "+c+" "+mc+" "+lv+" "+lvo);
         if (P.C[c] != null){                            
         P.C[c].MC[mc].N[ni].Kind=b[i*4+0]+(b[i*4+1]<<8);
         }                                               
                                                         
        }                                                
      } catch (IOException ex) {                         
        ex.printStackTrace();                            
      }                                                  
      Instant tve = Instant.now();                       
      long t = Duration.between(tvs, tve).toMillis();    
      System.out.println("  MSEC INIT:"+ t);             
    }else{                                               
      System.out.println("  parameter?");                
    }                                                    
    return P;                                            
  }                                                      
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
                                                         
  public static void iterate(Patch P){                   
    System.out.println("  iterating");                   
  }                                                      
                                                         
                                                         
                                                         
  public static void  summarize(Patch P){                
    System.out.println("  summarizing");                 
                                                         
  }                                                      
                                                         
                                                         
  public static void main (String[] args)    {           
    Patch P;                                             
    System.out.println(" Brainiac");                     
    P=initialize(args);                                  
    iterate(P);                                          
    summarize(P);                                        
  }                                                      
                                                         
                                                         
                                                         
                                                         
}                                                        
                                                         
                                                         
                                                         
                                                         
